SCIM integration topic (#5139)
pulumi/docs: content/docs/guides/scim/okta.md

---
title: Configuring SCIM in Okta
meta_desc: This page describes how to support SCIM 2.0 functionality between Pulumi and Okta.
menu:
    userguides:
        parent: scim
        weight: 1
---

> This document covers steps to help you configure automatic provisioning/deprovisioning of your users in Pulumi using SCIM 2.0.

Reviewer: ```suggestion
This document covers steps to help you configure automatic provisioning/deprovisioning of your users and groups in Pulumi using SCIM 2.0.
```

Please note that not all SCIM features are supported. For more information, see [Known Limitations]({{< relref "#known-limitations" >}}).

## Prerequisites

* Your organization must already be configured to use [SAML SSO]({{< relref "/docs/guides/saml/okta" >}}) with Pulumi.
* You must be an admin of your organization.
* (Optional) It is highly recommended that you have more than one admin for your org.

## Enabling SCIM For Your Okta Org

> In order to configure SCIM for your Okta tenant, email support@okta.com requesting them to enable `SCIM_PROVISIONING` for your Okta org.

Reviewer: ```suggestion
In order to configure SCIM for your Okta organization, you may need to email support@okta.com and request that they enable `SCIM_PROVISIONING` for the organization.
```

> Once the feature is enabled, navigate to the Pulumi SAML app configured in your Okta org from the **Classic UI**. Under the **General** tab, in the **App Settings** dialog box, locate the **Provisioning** section and select **SCIM**. Click **Save** to save your changes and close the **App Settings** dialog box.

Reviewer: This Okta doc has some decent screenshots (that they, not us, have the burden to update ;)) - would it be worth a link?

https://support.okta.com/help/s/article/SCIM-Provisioning-Enabled-in-Custom-App?language=en_US

Author: Agreed. Added link as part of https://github.com/pulumi/docs/pull/5139/commits/3b7f1b843826a1455935c066cda6e5bfbf01fb11.

1. Click **Next** on the first step.
2. On Step 2, in the **General** section, set the application username to **Okta username prefix**.
3. In the **Attribute Statements** section, click **Add Another** and set the **Name** of the property to `email` and the value to be source from `user.email`. The ATTRIBUTE STATEMENTS should now be as follows:

    | Name | Name Format | Value |
    | --- | --- | --- |
    | firstName | Basic | user.firstName |
    | lastName | Basic | user.lastName |
    | email | Basic | user.email |

4. Click **Next** then click **Finish**.

## Configuring the SCIM Connector

To configure the SCIM connector, click the **Provisioning** tab, and then select the **Integration** option in the left nav. Click **Edit**. Fill out the form as follows:

* SCIM connector base URL: `https://api.pulumi.com/scim/v2/{orgName}`, where `{orgName}` must be replaced with your organization’s login name (not display name). If you do not know this, navigate to your SAML settings and look at the SSO URL. It will have your organization’s login name in the URL.
* Supported provisioning actions: check **Push New Users** and **Push Profile Updates**.
* Authentication Mode: HTTP Header

> * For **HTTP HEADER**, you will used a token from the [Pulumi Console](https://app.pulumi.com) as the authorization bearer token. To generate a token, navigate to your org in the Pulumi Console, click on the **Settings** tab, and then click **SAML SSO**. Scroll down to the **SCIM** section and generate a new token if you have never generated one for your org, or regenerate it if you have already done so in the past.

Reviewer 2: ```suggestion
* For **HTTP HEADER**, you will use a token from the [Pulumi Console](https://app.pulumi.com) as the authorization bearer token. To generate a token, navigate to your org in the Pulumi Console, click on the **Settings** tab, and then click **SAML SSO**. Scroll down to the **SCIM** section and generate a new token if you have never generated one for your org, or regenerate it if you have already done so in the past.
```

    {{% notes "info" %}}

> Once you generate the token, please save it securely. The Pulumi Console will not show you the token again and there is no way to retrieve it again. SCIM token can only be regenerated once it has been generated already. This means you will invalidate any previous SCIM tokens for your org by regenerating it.

Reviewer: ```suggestion
Once you generate the token, please save it securely. Neither the Pulumi Console nor Pulumi support can retrieve a token once it's been initially generated. If you lose the SCIM token and need it again, you'll have to generate a new token, which invalidates any previous tokens for your Pulumi organization.
```

    {{% /notes %}}

* Paste the token from the Pulumi Console into the Okta SCIM configuration for **HTTP HEADER**.

The following shows how your SCIM connection info should be filled in.

![Okta SCIM provisioning](/images/docs/guides/scim/okta-scim-provisioning.png)

## Enabling SCIM Provisioning Actions

> There are a few provisioning actions you need to enable, which Okta will con on the Pulumi service.

Reviewer 2: There seems to be something missing in the second part of this sentence.

Reviewer: ```suggestion
To configure the actions that Okta will send to the Pulumi Console, you need to enable them.
```

1. Under the **Provisioning** tab, click **Edit**.
2. While you are in the **To App** section, check **Enable** on the following:

    * Create Users
    * Update User Attributes
    * Deactivate Users
3. Click **Save**.

## Adjusting the Okta to Pulumi Attribute Mappings

Now that the SCIM connector knows how to connect to Pulumi, you need to adjust the attributes that Okta will send to Pulumi and what it might receive back from Pulumi. To do that, perform the following steps:

1. Under **Provisioning** > **To App**, click **Go to Profile Editor**.
2. Click **Mappings** to open the Attribute Editor dialog box.
3. Un-map all but the following attributes:

>     * firstName
>     * lastName

Reviewer 2: Okta uses slightly different terminology.

```suggestion
    * givenName
    * familyName
```

    * displayName
    * email

4. Click on the **Okta User To …** tab and un-map all but the following attributes:

>     * firstName
>     * lastName

Reviewer 2: ```suggestion
    * givenName
    * familyName
```

    * displayName
    * email

5. Click **Save Mappings**.
6. Click **Apply Updates**.

## Setting up Group Provisioning

> The Pulumi console supports the provisioning of teams within your organization using SCIM. This is done by mapping the groups you have created using SCIM to create teams within your organization in the Pulumi console. Setting this up allows you to manage your team’s membership solely from the Okta side.

Reviewer: ```suggestion
The Pulumi console supports the provisioning of teams within your organization using SCIM. This is done by mapping the groups you have created using SCIM to create teams within your organization in the Pulumi console. Setting this up allows you to manage your teams' memberships solely in Okta.
```

> To do this, you will enable Push Groups as a supported provisioning action under the **Provisioning** settings and then specify which groups you would like to push through the following steps:

Reviewer: ```suggestion
To set this up, you need to enable Push Groups as a supported provisioning action under the **Provisioning** settings and then specify which groups you would like to push. To do that, performing the following steps:
```

> To set this up, you need to enable Push Groups as a supported provisioning action under the **Provisioning** settings and then specify which groups you would like to push. To do that, performing the following steps:

Reviewer 3: ```suggestion
To set this up, you need to enable Push Groups as a supported provisioning action under the **Provisioning** settings and then specify which groups you would like to push. To do that, perform the following steps:
```

1. Navigate to the application that you created above by clicking the on the **Applications** tab and then selecting the application.

    ![Okta application selection.](/images/docs/guides/scim/okta-application-selection-scim.png)

2. Click on **Provisioning** tab, then click on the integrations menu option located on the side navigation.
3. Check the **Push Groups** box under **Supported provisioning actions**.
4. Select the **Push Groups** tab.
5. Select the **Push Groups** drop down menu. Here you will have the option to find groups either by name or rule.

    ![Push Groups drop down menu in Okta](/images/docs/guides/scim/okta-pushgroups-menu-scim.png)

6. Select **Find groups by name**. This will take you to a screen where you can search groups by the group name. Once you find the group(s) that you are looking for, click **Save**.

    You will notice there is a checkbox labeled **Push group memberships immediately**. This will also add all the members that are part of your group to the Pulumi team after your selection has been saved.

7. After clicking **Save**, you should see that your group was pushed and it should say **Active**.

>     Note: if there are members in a group that are not yet assigned to the Pulumi Console application in Okta, they will not be added to the team in the Pulumi Console. Ensure that all members in the group have been assigned to the application before pushing the group.

Reviewer 3: Should this note contain the notes styling similar to how it is done on line 51?

Author: Yup, good catch. Included in https://github.com/pulumi/docs/pull/5139/commits/73b0e75c5d641c053a4fa585ceeb0101abb21dd7

>     One thing to note is that if there are members in your group that are not yet assigned to the Pulumi Console application, they will not be added to the team on the Pulumi end. The Pulumi Console will make a best effort to add the members being requested. If any of them are not assigned to the Pulumi Console application or their user has not yet been pushed we will simply skip over adding that member.

Reviewer: ```suggestion
    Note: if there are members in a group that are not yet assigned to the Pulumi Console application in Okta, they will not be added to the team in the Pulumi Console. Ensure that all members in the group have been assigned to the application before pushing the group.
```

## Verifying Group Provisioning

To confirm that the groups were provisioned correctly, log into the Pulumi Console and select the **TEAMS** tab where you should see the team listed.

![Pulumi console showing current teams.](/images/docs/guides/scim/console.png)

Click on that group and verify its membership. Note that the team’s membership cannot be altered within the Pulumi console. If any membership changes are needed, they must be done within Okta. This allows the membership to be managed using Okta so your teams on the Pulumi side will always mirror the groups you have configured there.

## Known Limitations

Here’s a list of the features that are not currently supported:

* Secondary emails
* Password sync
* Bulk importing

> * Subgroups (groups composed of other groups, only groups composed of users are supported)

Reviewer 3: I recently discovered that Okta does not support subgroups either so I think we can remove this item since this doc is Okta specific.

## Changing the SAML SSO Settings For SCIM

As part of the setup process, you need to adjust some of the parameters sent by a SAML SSO request. To make the changes, navigate to the **SAML Settings** section, click **Edit**, and make the following changes:

1. Click **Next** on the first step.
2. On Step 2, in the **General** section, set the application username to **Okta username prefix**.
3. In the **Attribute Statements** section, click **Add Another** and set the **Name** of the property to `email` and the value to be source from `user.email`. The ATTRIBUTE STATEMENTS should now be as follows:

    | Name | Name Format | Value |
    | --- | --- | --- |
    | firstName | Basic | user.firstName |
    | lastName | Basic | user.lastName |
    | email | Basic | user.email |

4. Click **Next** then click **Finish**.

## Configuring the SCIM Connector

To configure the SCIM connector, click the **Provisioning** tab, and then select the **Integration** option in the left nav. Click **Edit**. Fill out the form as follows:

> * SCIM connector base URL: `https://api.pulumi.com/scim/v2/{orgName}`, where `{orgName}` must be replaced with your organization’s login name (not display name). If you do not know this, navigate to your SAML settings and look at the SSO URL. It will have your organization’s login name in the URL.

Reviewer 2: There is another setting on this page called "Unique identifier field for users" and it should be set to `userName`.

Author: Fixed in https://github.com/pulumi/docs/pull/5139/commits/77514aeab0a1553b7045b40df59bef70c4a62513.

> * Supported provisioning actions: check **Push New Users** and **Push Profile Updates**.

Reviewer 2: Should we also tell users to enable "Push Groups" here if they are planning to use that feature as well rather than having to come back here later anyway?

Author: @praneetloke I think it's best to provide a link to the group provisioning section here given the extra look-up steps that seem to be involved.

---
title: Configuring SCIM in Okta
meta_desc: This page describes how to support SCIM 2.0 functionality between Pulumi and Okta.
menu:
    userguides:
        parent: scim
        weight: 1
---

> This document covers steps to help you configure automatic provisioning/deprovisioning of your users and groups in Pulumi using SCIM 2.0.

Reviewer: ```suggestion
This document outlines the steps required to help you configure automatic provisioning/deprovisioning of your users and groups in Pulumi using SCIM 2.0.
```

> Please note that not all SCIM features are supported. For more information, see [Known Limitations]({{< relref "#known-limitations" >}}).

Reviewer: ```suggestion
Please note that some advanced SCIM features aren't supported yet. For more information, see [Known Limitations]({{< relref "#known-limitations" >}}).
```

## Prerequisites

> * Your organization must already be configured to use [SAML SSO]({{< relref "/docs/guides/saml/okta" >}}) with Pulumi.
> * You must be an admin of your organization.
> * (Optional) It is highly recommended that you have more than one admin for your org.

Reviewer: ```suggestion
* Your organization must already be configured to use [SAML SSO]({{< relref "/docs/guides/saml/aad" >}}) with Pulumi.
* You must be an admin of your Pulumi organization.
* (Optional, but highly recommended) You should have more than one admin for your Pulumi organization.
```

7. After clicking **Save**, you should see that your group was pushed and it should say **Active**.

    {{% notes "info" %}}
If there are members in a group that are not yet assigned to the Pulumi Console application in Okta, they will not be added to the team in the Pulumi Console. Ensure that all members in the group have been assigned to the application before pushing the group.
    {{% /notes %}}

## Verifying Group Provisioning

To confirm that the groups were provisioned correctly, log into the Pulumi Console and select the **TEAMS** tab where you should see the team listed.

![Pulumi console showing current teams.](/images/docs/guides/scim/console.png)

Click on that group and verify its membership. Note that the team’s membership cannot be altered within the Pulumi console. If any membership changes are needed, they must be done within Okta. This allows the membership to be managed using Okta so your teams on the Pulumi side will always mirror the groups you have configured there.

## Known Limitations

> Here’s a list of the features that are not currently supported:

Reviewer: ```suggestion
Some SCIM features are currently not supported:
```
